### PR TITLE
Added Pack Hunter Mutation

### PR DESCRIPTION
**Pull request by pehamm**
#### Summary
Content "Added Pack Hunter mutation for post-threshold lupines."

#### Purpose of change

When playing a volunteer mutant, I evolved into the lupine branch, crossed the threshold, and then noticed that the lupine tree only has the predator mutations as their post-threshold. Thus makes their post-threshold mutations a strict subset of the feline and beast trees; there is no real reason to cross the threshold and it offers nothing specific to wolves.

#### Describe the solution

I added a mutation called "Pack Hunter" that evolves from "Predator" and retains the benefits of Predator while removing the Intelligence penalty. My reasoning is that the most distinguishing feature of a wolves hunting style compared to cats (lions nonwithstanding) or bears is cooperation and social interaction, which demands a certain level of reasoning and planning ability. Just like wolves learned to cooperate with humans and became dogs, a lupine player may see some value in maintaining some relations with humans.

#### Describe alternatives you've considered

Technically, I considered implementing this as a seperate mutation requiring "Predator" that simply re-adds the intelligence lost, but I could not find a description that did not directly contradict the one from "Predator", concerning how the players critical thinking suffers. Thanks to anothersimulacrum for pointing out that I can do all of this in JSON.

#### Testing

Started a character, spawned a pile of lupine mutagen and serum. Crossed the threshold, gained "Predator" and lost an intelligence point, then gained "Pack Hunter" and regained my Intelligence Point.

#### Additional context

Unsure about the description. I have

> "Puny as they may seem, a hunter needs a pack, and some of those hairless bipeds may yet prove some worth. You continue to gain combat skills at an elevated rate, and maintain a level of critical thinking that supports cooperation."

trying to keep with the tone of the Predator mutations, but I am open to any and all feedback.

Currently, checks fail due to
`Error: (continued from above) ERROR : src/regional_settings.cpp:622 [void check_region_settings()] Invalid map extras for region "default", extras "default".  Extras "mx_portal" and "mx_portal_in" are listed, but all have zero weight.`
which should not be affected at at by this PR. I assume it is a random fluctuation, as other PR seem to pass this fine.

### CODE REVIEW

**Comment by pehamm** [comment]
Blanked out and forgot to lint everything, will fix the issues locally.

File: src/character.cpp [hunk 1]
```diff
@@ -2347,6 +2347,10 @@ void Character::practice( const skill_id &id, int amount, int cap, bool suppress
         amount *= 2;
     }
 
+    if( has_trait_flag( "PACK_HUNTER" ) && skill.is_combat_skill() ) {
+        amount *= 2;
+    }
+
```
Comment (anothersimulacrum) [inline src/character.cpp:2353]: This flag should not be used here. Either add the `PRED3` flag to your mutation, or add one for what it does in common with this.

File: src/character.cpp [hunk 2]
```diff
@@ -4257,6 +4262,7 @@ void Character::do_skill_rust()
         if( aSkill.is_combat_skill() &&
             ( ( has_trait_flag( PRED2 ) && calendar::once_every( 8_hours ) ) ||
               ( has_trait_flag( PRED3 ) && calendar::once_every( 4_hours ) ) ||
+              ( has_trait_flag( PACK_HUNTER ) && calendar::once_every( 4_hours ) ) ||
```
Comment (anothersimulacrum) [inline src/character.cpp:4265]: Ditto

File: src/consumption.cpp [hunk 3]
```diff
@@ -1152,13 +1152,14 @@ void Character::modify_morale( item &food, const int nutr )
...
         const bool predator = has_trait_flag( "PRED3" );
+        const bool pack_hunter = has_trait_flag( "PACK_HUNTER" );
         const bool apex_predator = has_trait_flag( "PRED4" );
         if( apex_predator ) {
             // Largest bonus, balances out to around +5 or +10. Some organs may still be negative.
             add_morale( MORALE_MEATARIAN, 20, 10 );
             add_msg_if_player( m_good,
                                _( "As you tear into the raw flesh, you feel satisfied with your meal." ) );
-        } else if( predator || hunter ) {
+        } else if( predator || hunter || pack_hunter ) {
```
Comment (anothersimulacrum) [inline src/consumption.cpp:1162]: All of these should really be one trait flag.

File: src/mondeath.cpp [hunk 4]
```diff
@@ -419,6 +419,7 @@ void mdeath::guilt( monster &z )
 
     Character &player_character = get_player_character();
     if( player_character.has_trait( trait_PSYCHOPATH ) || player_character.has_trait_flag( "PRED3" ) ||
+        player_character.has_trait_flag( "PACK_HUNTER" ) ||
```
Comment (anothersimulacrum) [inline src/mondeath.cpp:422]: Again, if this does the same thing as PRED3, use a common flag. PRED4 (and all these traits) should use one generic flag for this check.

**Comment by anothersimulacrum** [comment]
This really shouldn't be requiring C++ changes, but apparently the people who moved these to trait flags didn't believe in being very generic (at least with the names).

**Comment by pehamm** [comment]
Agree, a lot of duplication in the C++ code. I worked on the assumption that each mutation has to use its own unique flag as this is what I saw with the PREDn mutations and I learned my lesson from tidying up code I do not fully understand yet. If I can use them more flexibly I will see if I can tidy this whole thing up, thank you for the input!

**Comment by pehamm** [comment]
You are right, I misunderstood how the flags worked, using the PRED3 flag works just as well.

Meanwhile, I messed up reverting my C++ changes as the files were changed on the master branch and I forgot to check for that. I just reverted all commits and recommited with just the new JSON.

File: src/character.cpp (hunk 1 above)
Comment (pehamm) [reply src/character.cpp]: Reverted all changes to the file, as they were not needed after all.

**Comment by pehamm** [comment]
Still trying to work out why one Travis build fails. If I read the log correctly, the code compiles fine, but fails a field test saying "Terrain should be irradiated under 5 minutes" that should in no way be affected by me adding a mutation in JSON as far as I can tell.

Will try to download Xenial tomorrow and rerun with the same setup Travis used.

**Comment by actual-nh** [comment]
> Still trying to work out why one Travis build fails. If I read the log correctly, the code compiles fine, but fails a field test saying "Terrain should be irradiated under 5 minutes" that should in no way be affected by me adding a mutation in JSON as far as I can tell.

The field tests sometimes give errors like that - it's random exactly how fast a field expands/contracts, so the tests have a large margin of error... but no margin (that would leave the test meaningful) is enough to always ensure against bad luck.

**Comment by pehamm** [comment]
> The field tests sometimes give errors like that - it's random exactly how fast a field expands/contracts, so the tests have a large margin of error... but no margin (that would leave the test meaningful) is enough to always ensure against bad luck.

Thank you for the insight. I thought fixing the seeds would make all of the tests nonrandom, but I am still learning how the code functions as a whole. Can I safely ignore the build failure then or do I have to get everything to pass before this has a chance to be merged?

**Comment by actual-nh** [comment]
> Thank you for the insight. I thought fixing the seeds would make all of the tests nonrandom, but I am still learning how the code functions as a whole. Can I safely ignore the build failure then or do I have to get everything to pass before this has a chance to be merged?

AFAIK, the seeds aren't fixed (tradeoff: better test coverage vs occasional false positives - the classic sensitivity vs specificity tradeoff). You can ignore (albeit perhaps comment on) failures that are not reasonably related to what you're altering ('reasonably' may differ in the eye of the beholder, admittedly; commenting on failures may help to convey this information).

**Review by anothersimulacrum — APPROVED** [review]
Looks good to me!

**Pull request merged**